Autofill support for deleting all passwords

amddg44 opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.

⚠️ If you're an external contributor, please file an issue first before working on a PR, as we can't guarantee that we will accept your changes if they haven't been discussed ahead of time. Thanks!
-->

Task/Issue URL: https://app.asana.com/0/1201462886803403/1206303310228833/f
Tech Design URL:
CC:

**Description**:
Give users the ability to delete all passwords they have saved

<!--
If at any point it isn't actively being worked on/ready for review/otherwise moving forward strongly consider closing it (or not opening it in the first place). If you decide not to close it, use Draft PR while work is still in progress or use `DO NOT MERGE` label to clarify the PRs state and comment with more information.
-->

**Steps to test this PR**:

**Delete all (multiple passwords saved)**

1. Go to the Passwords screen & save 2 or more passwords
2. Tap `Edit` and confirm the toolbar appears as per the figma design
3. Confirm the top section which contains the row `Save and autofill passwords` is removed
4. Confirm the toobar label has the correct password count
5. Tap `Delete All`
6. Confirm the alert dialog text includes the correct password count
7. If sync is enabled, confirm it is mentioned in the alert message or if sync is not enabled, confirm it is not mentioned in the alert message
8. Follow the prompts and confirm you are prompted to authenticate in order to delete all
9. Cancel / fail the authentication, making sure the delele-all flow is cancelled
10. Repeat and this time authenticate successfully
11. Confirm all passwords disappear from the UI
12. Confirm the toast prompt appears with the correct password count & an `Undo` button
13. Tap `Undo` and confirm the passwords re-appear
14. Back out of the screen and come back, confirming the passwords still appear correctly
15. Repeat but this time don't tap `Undo`.
16. Back out of the screen and come back, confirming there are no passwords saved

**Delete one password**

1. Tap `Edit` and delete a single password
2. Confirm the toolbar password count updates correctly

**Delete all (one password saved)**

The copy is slightly different in the case of the user having just a single password saved:
1. Ensure only a single password is saved
2. Tap `Delete All` and confirm the alert title mentions "this password" instead of the password count
3. Follow the complete `Delete All` flow and confirm the singular is used in all the copy in reference to passwords / accounts

**Password details screen**

1. Tap a password to see its details
2. Tap the `Delete` button and confirm the title & message are the same as the `Delete all (one password saved)` alert 
3. If sync is enabled, confirm it is mentioned in the alert message or if sync is not enabled, confirm it is not mentioned in the alert message

**Sync**

1. With Sync enabled, confirm the passwords are deleted on the other devices
2. Depending on whether or not you were testing with Sync enabled / disabled, turn it off / on and confirm the `Delete all` alert copy doesn't / does mention sync in the alert message

**Toolbar behaviour**

1. Confirm the toolbar only appears when in `Edit` mode and not in any of these cases:
- when table is empty
- when searching
- when accessing password details
- when screen is locked for authentication
2. Access the Passwords screen via Settings (not from the menu overflow)
3. Tap `Edit` and back out of the Passwords screen
4. Confirm the toolbar is correctly removed

**Authentication**

1. Repeat the `Delete all` flow but this time minimise the app at each stage
2. Confirm authentication is always required to view the list of passwords when returning to the app


**Save a never prompt site**

1. Visit `https://fill.dev/form/login-simple` and enter some credentials
2. Tap `Never Ask for this Site`
3. Go to the Passwords screen and confirm the row `Reset Excluded Sites` is now present in the top section
4. Delete all passwords & confirm this row disappears after the toast disappears 

**Additional smoke tests**

1. Lightly test search & confirm everything still works correctly


<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
-->

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [x] Portrait
* [x] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16

**Theme Testing**:

* [x] Light theme
* [x] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

amddg44 committed:
New prompt for explicit auth of delete all

amddg44 committed:
new copy

amddg44 committed:
New files added to .pbxproj

amddg44 committed:
point to BSK branch

amddg44 committed:
min height for prompt added

amddg44 committed:
support added to clearAll and deleteAll accounts

amddg44 committed:
Toolbar with delete all button

amddg44 committed:
Delete all button now triggers the AuthConfirmationPrompt

amddg44 committed:
Update viewModel to update sections based on whether or not tableView isEditing

amddg44 committed:
Removed unnecessary code

amddg44 committed:
Corrected indentation

amddg44 committed:
Delegate added to notify when user is authenticating so that notification observers can be updated

amddg44 committed:
Notification observers cleaned up, with willEnterForegroundNotification replaced by correct didBecomeActiveNotification + functions given better naming

amddg44 committed:
Update observers while user is authenticating the delete all action

amddg44 committed:
Updated to use same copy as delete all

amddg44 committed:
Fix for toolbar label not updating when a single item deleted

amddg44 committed:
fix linter issue

amddg44 committed:
Fix tests

amddg44 committed:
Extra space character removed from string

amddg44 committed:
Remove toolbar when backing out via Settings while in edit mode

amddg44 committed:
Add suggestions section back to edit mode

amddg44 committed:
Update toolbar to use systemRed

amddg44 committed:
Fine tuning toolbar to only display in viewState .showItems

amddg44 committed:
Delete message content dynamic based on users sync status

amddg44 committed:
Merge remote-tracking branch 'origin/main' into autofill/anya-delete-all-ios

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
Update BSK commit ref

aataraxiaa commented on DuckDuckGo/AuthConfirmationPromptViewController.swift:
```diff
@@ -0,0 +1,106 @@
+import UIKit
+import SwiftUI
+
+class AuthConfirmationPromptViewController: UIViewController {
```
Nitpick: Maybe could mark this `final`? 

```suggestion
final class AuthConfirmationPromptViewController: UIViewController {
```

aataraxiaa commented on DuckDuckGo/AuthConfirmationPromptViewController.swift:
```diff
@@ -0,0 +1,106 @@
+    
+    private var viewModel: AuthConfirmationPromptViewModel?
+    
+    internal init(didBeginAuthenticating: @escaping () -> Void,
```
Curious: Why do we need to mark this `init` internal?

amddg44 replied:
Yup, no need for this to be marked internal; updated

aataraxiaa commented on DuckDuckGo/AuthConfirmationPromptViewModel.swift:
```diff
@@ -0,0 +1,60 @@
+    func authConfirmationPromptViewModelDidResizeContent(_ viewModel: AuthConfirmationPromptViewModel, contentHeight: CGFloat)
+}
+
+class AuthConfirmationPromptViewModel: ObservableObject {
```
Nitpick: Could maybe mark this `final`?

amddg44 committed:
Unit tests added for view model changes

aataraxiaa commented on DuckDuckGo/AutofillLoginDetailsViewController.swift:
```diff
@@ -38,6 +39,7 @@ class AutofillLoginDetailsViewController: UIViewController {
     private let viewModel: AutofillLoginDetailsViewModel
     private var cancellables: Set<AnyCancellable> = []
     private var authenticator = AutofillLoginListAuthenticator(reason: UserText.autofillLoginListAuthenticationReason)
+    private let syncService: DDGSyncing
```
Maybe we can remove this? I think it is only passed into the ViewModel and then not referenced again.

amddg44 replied:
Good point! Done 🙂

aataraxiaa reviewed (approved):
Verified scenarios and left some very minor comments. LGTM @amddg44. 🥳

amddg44 committed:
Bump BSK revision

amddg44 committed:
Merge remote-tracking branch 'origin/main' into autofill/anya-delete-all-ios

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
Address PR feedback

amddg44 committed:
Linter fix

amddg44 committed:
Minor copy updates from review

amddg44 committed:
Merge remote-tracking branch 'origin/main' into autofill/anya-delete-all-ios

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
Update BSK revision

amddg44 committed:
Merge remote-tracking branch 'origin/main' into autofill/anya-delete-all-ios

amddg44 committed:
Merge remote-tracking branch 'origin/main' into autofill/anya-delete-all-ios

amddg44 committed:
Update BSK revision

amddg44 committed:
Translations for deleting all passwords (#2533)

Task/Issue URL: https://app.asana.com/0/1201462886803403/1206303310228833/f
Tech Design URL:
CC:

Description:
Adds translations to autofill support for deleting all passwords

amddg44 committed:
Merge remote-tracking branch 'origin/main' into autofill/anya-delete-all-ios

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
Update BSK revision

amddg44 committed:
Set to BSK release 114.3.0
